### PR TITLE
[wip]: fsspec remote code cache

### PR DESCRIPTION
**Pull request by msaroufim**
The applications for this PR are around reducing cold start times
1. Going from training to inference in case you're using the same machine
2. Distributed
3. CI where the majority of some code does not change and you'd like to reduce test time

fsspec is amazing and can support the widest possible amount of file systems in my tests i mocked s3 and local paths but more setups could work, it's already a pytorch dependency https://download.pytorch.org/whl/nightly/fsspec/ 

In case test is skipped in CI

```
(sourcetorch) ubuntu@ip-172-31-1-136:~/pytorch/test/inductor$ pytest -s test_cache.py
============================================================================ test session starts ============================================================================
platform linux -- Python 3.10.10, pytest-7.3.1, pluggy-1.0.0
rootdir: /home/ubuntu/pytorch
configfile: pytest.ini
plugins: hypothesis-6.75.1
collected 4 items                                                                                                                                                           

test_cache.py ....

============================================================================= 4 passed in 2.74s =============================================================================
(sourcetorch) ubuntu@ip-172-31-1-136:~/pytorch/test/inductor$ 
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
havent reviewed code yet

File: test/inductor/test_cache.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import pkgutil
+import shutil
+import unittest
+import unittest.mock as mock
+
+import torch._inductor.codecache as codecache
+
+requires_fsspec = functools.partial(
+    unittest.skipIf, not pkgutil.find_loader("fsspec"), "requires fsspec"
+)
```
[the remark this answers is not in the document]
Comment (msaroufim) [reply test/inductor/test_cache.py:12]: It is included yeah https://download.pytorch.org/whl/nightly/fsspec/

[the remark this answers is not in the document]
Comment (msaroufim) [reply test/inductor/test_cache.py:12]: So for the review what's tripping me up right now is which parts of the codebase need to be fsspec aware, after seeing all the usages of cache_dir it seems like I need to make some pretty big changes - see for example the `UniversalPath` thing which seems like I should probably upstraem to fsspec instead

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -305,13 +395,30 @@ def check_cache(cache, callback=None) -> bool:
...
 
+
 def get_lock_dir() -> str:
-    lock_dir = os.path.join(cache_dir(), "locks")
-    if not os.path.exists(lock_dir):
-        os.makedirs(lock_dir, exist_ok=True)
+    lock_dir = join_paths(cache_dir(), "locks")
+    
+    # Check if the path is an fsspec path
+    if "://" in lock_dir:
```
Comment (msaroufim) [inline torch/_inductor/codecache.py:403]: could probably have a helper function for this pattern

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -460,12 +567,28 @@ def cpp_compiler_search(search: str) -> str:
     raise exc.InvalidCxxCompiler()
 
 
+import fsspec
+
 def install_gcc_via_conda() -> str:
```
Comment (msaroufim) [inline torch/_inductor/codecache.py:572]: add the try except pattern

File: torch/_inductor/codecache.py [hunk 4]
```diff
@@ -460,12 +567,28 @@ def cpp_compiler_search(search: str) -> str:
     raise exc.InvalidCxxCompiler()
 
 
+import fsspec
```
Comment (msaroufim) [inline torch/_inductor/codecache.py:570]: remove

**Pull request closed without merging**